Libplanet/Blockchain/BlockChain.Swap.cs

```diff
@@ -0,0 +1,462 @@
+                throw new InvalidGenesisBlockException(Genesis.Hash, other.Genesis.Hash, msg);
+            }
+
+            _logger.Debug(
+                "The branchpoint is #{BranchpointIndex} {BranchpointHash}.",
+                branchpoint.Index,
+                branchpoint
+            );
+
+            _rwlock.EnterUpgradeableReadLock();
+            try
+            {
+                Block<T> oldTip = Tip ?? Genesis, newTip = other.Tip ?? other.Genesis;
+
+                ImmutableList<Block<T>> rewindPath =
+                    GetRewindPath(this, branchpoint.Hash);
+                ImmutableList<Block<T>> fastForwardPath =
+                    GetFastForwardPath(other, branchpoint.Hash);
+
+                // If there is no rewind, it is not considered as a reorg.
+                bool reorg = rewindPath.Count > 0;
+
+                _rwlock.EnterWriteLock();
+                try
+                {
+                    IEnumerable<Transaction<T>>
+                        GetTxsWithRange(BlockChain<T> chain, Block<T> start, Block<T> end)
+                        => Enumerable
+                            .Range((int)start.Index + 1, (int)(end.Index - start.Index))
+                            .SelectMany(x => chain[x].Transactions);
+
+                    // It assumes reorg is small size. If it was big, this may be heavy task.
+                    ImmutableHashSet<Transaction<T>> unstagedTxs =
+                        GetTxsWithRange(this, branchpoint, Tip).ToImmutableHashSet();
+                    ImmutableHashSet<Transaction<T>> stageTxs =
+                        GetTxsWithRange(other, branchpoint, other.Tip).ToImmutableHashSet();
+                    ImmutableHashSet<Transaction<T>> restageTxs = unstagedTxs.Except(stageTxs);
+                    foreach (Transaction<T> restageTx in restageTxs)
+                    {
+                        StagePolicy.Stage(this, restageTx);
+                    }
+
+                    Guid obsoleteId = Id;
+                    Id = other.Id;
+                    Store.SetCanonicalChainId(Id);
+                    _blocks = new BlockSet<T>(Policy.GetHashAlgorithm, Store);
+                    TipChanged?.Invoke(this, (oldTip, newTip));
+
+                    Store.DeleteChainId(obsoleteId);
+                }
+                finally
+                {
+                    _rwlock.ExitWriteLock();
+                }
+
+                renderSwap = () => RenderSwap(
+                    render: render,
+                    oldTip: oldTip,
+                    newTip: newTip,
+                    branchpoint: branchpoint,
+                    rewindPath: rewindPath,
+                    fastForwardPath: fastForwardPath,
+                    stateCompleters: completers);
+            }
+            finally
+            {
+                _rwlock.ExitUpgradeableReadLock();
+            }
+
+            return renderSwap;
+        }
+
+        internal void RenderSwap(
+            bool render,
```

Comment:
How about removing this parameter too?

Reply:
There are unnecessary parameters still left in the code here and there just for the sake of consistency, but `render` is not one of them. If it is to be removed, it should be from `Swap()` not `RenderSwap()`. More logical implementation would be to return `Action<bool>` instead of `Action`, but either way, I'd say `render` is necessary in `RenderSwap()`.

Reply:
Although I agree the design isn't optimal at this point, if you don't mind too much, I think I'd like to deal with this issue separately. See #1417. Another minor reason `render` is still left there is for consistency, without too much thought into how to streamline the rendering processes.